### DISCUSSION TITLE
[FEATURE] Service Output Register

### DISCUSSION THREAD

**Discussion by tmob03** [ArgosOfIthica/scalu#48] [General]
**Is your feature request related to a problem? Please describe.**
Like with `std.output`, we are already making registers for outputs of services, but these could be abstracted from end users, not needing to define these themselves per script.

**Describe the solution you'd like**
1-4(or so, maybe even definable with a compiler statement at start, or even a smart compiler, to see how many nested services there are at once?) registers which hold a service's output, eliminating needing to manually assign a variable at the end of a service. Possible syntax:

```
service add {
    return input1 + input2  /* implicitly equal to currentSandbox.output = input1 + input2 */
}

service some_other_service {
    input1 = 70
    input2 = 125
    @add(answer) /* sets variable answer to the output of add /*
}
```

This syntax may not be great if you consider ever adding abstracted service inputs, but in my opinion those go a little far for performance requirements, unless the compiler can determine the max amount of input registers needed on compile. In that case I'd say just pretty much normal function calling syntax in other languages
`answer = @add(input1, input2) /* answer set to the output of add service, passing input1 and input2 */`

**Describe alternatives you've considered**
In addition to the `return` keyword, there could be a `hard_return` keyword, or just make return by default do this, although it leaves overhead doing so. `hard_return` would be like a normal programming language return statement, ending the function where it's met. It can return a number just like `return`, or it could be by itself and simply force the service to terminate. An example of termination is below, mostly taken from my tic tac toe game, and what its equivalent scalu code is:

Suggestion:
```
service input_4 {
    if (space4 != 0) {
        [echo This space is filled, try again]
        @example_service /* game execution goes to this service, them comes back to `input_4` */
        hard_return /* returning a value of null, simply forcing the service to terminate */
    }
       [echo You placed an O on the fourth space]
	rng = rng + 2
	space4 = 1
	@render_board
	@check_win
	if (win == 0) {
	@ai_turn
}
```

Equivalent:
```
service input_4 {
	if (space4 != 0) {
		[echo This space is filled, try again]
                @example_service
	} else {
		[echo You placed an O on the fourth space]
		rng = rng + 2
		space4 = 1
		@render_board
		@check_win
		if (win == 0) {
		@ai_turn
		}
	}
}
```
Notice this is done by using an else statement, so when the game script execution comes back to finish this service's alias, it sees its an else statement, thus it doesn't run. This is why I say to have both `return` and `hard_return`, return doesn't cause this else checking, as it won't terminate the current service. 

**Additional context**
Not sure how viable all this is, but it would certainly make my life a lot easier!

**Comment by ArgosOfIthaca** [comment]
Just some thoughts on this.

This sort of comes back to why there are "services" and not "functions". I can have input/output with services and not have recursion, or I can not have input/output and have recursion. To do both, I need to implement stack semantics to keep track of the local variables on each frame of an unbounded stack, which would be wildly inefficient with our runtime. Given recursion gives users more actual power, and our computational construct, a service, usually expects to be triggered by a construct that also can't provide arguments (events), scalu just throws the issue of input and output back at the user to deal with. Not saying anything like this is impossible, but that's the logic that's guided development to this point.

I've considered a similar construct to `hard_return` where you halt all execution; I've thought about adding a small runtime engine that stops the possibility of infinite looping yourself (at least in a scalu script).

To achieve a `hard_return` like concept, you have to acknowledge that the assembly metaphor scalu uses breaks down; we cannot "goto" a different instruction than intended, as we don't really have an instruction pointer to simply change, but rather we have to expand and execute all the aliases in the command queue pipeline for the current game frame. The two paths are we can either delete the alias definitions of what's in the pipeline and restore them later, or we rewrite the control flow, similar to what you did in your example, but also for any situation where its nested arbitrarily. The efficacy of zero cost control flow rewriting was actually already answered by [structured programming](https://dl.acm.org/doi/10.1145/356635.356640) results in the 70's in the negative. That leaves deleting definitions, which is a lot more theoretically simple and superficially lower cost, but might negatively affect minifier optimizations based on aliasing once. I'll look deeper into it at some point.
